cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/RegularExpression.java

```diff
@@ -20,34 +23,52 @@
         this.expressionRegexp = expressionRegexp;
         this.parameterTypeRegistry = parameterTypeRegistry;
-        treeRegexp = new TreeRegexp(expressionRegexp);
+        this.treeRegexp = new TreeRegexp(expressionRegexp);
     }
 
     @Override
-    public List<Argument<?>> match(String text) {
-        List<ParameterType<?>> parameterTypes = new ArrayList<>();
+    public List<Argument<?>> match(String text, Type... typeHints) {
+        final ObjectMapper objectMapper = parameterTypeRegistry.getObjectMapper();
+        final List<ParameterType<?>> parameterTypes = new ArrayList<>();
+        int typeHintIndex = 0;
         for (GroupBuilder groupBuilder : treeRegexp.getGroupBuilder().getChildren()) {
-            String parameterTypeRegexp = groupBuilder.getSource();
+            final String parameterTypeRegexp = groupBuilder.getSource();
+            final Type typeHint = typeHintIndex < typeHints.length ? typeHints[typeHintIndex++] : Object.class;
 
             ParameterType<?> parameterType = parameterTypeRegistry.lookupByRegexp(parameterTypeRegexp, expressionRegexp, text);
-            if (parameterType == null) parameterType = new ParameterType<>(
-                    null,
-                    parameterTypeRegexp,
-                    String.class,
-                    new Transformer<String>() {
-                        @Override
-                        public String transform(String arg) {
-                            return arg;
-                        }
-                    }
-            );
+
+            if (parameterType == null) {
+                parameterType = createAnonymousParameterType(parameterTypeRegexp);
+            }
+
+            // Either from createAnonymousParameterType or lookupByRegexp
+            if (parameterType.isAnonymous()) {
+                Transformer<Object> transformer = new ObjectMapperTransformer(objectMapper, typeHint);
+                parameterType = parameterType.deAnonymize(typeHint, transformer);
+            }
+
+            // Use hint for validation only
+            // TODO: Handle types that are not classes - needs tests
+            Class<?> paramClass = (Class) parameterType.getType();
+            Class<?> hintClass = (Class) typeHint;
+            if (!hintClass.isAssignableFrom(paramClass)) {
+                throw new CucumberExpressionException(String.format(
+                        // TODO: Tell users what to do if they attempt to convert \d+ into Float.
+                        "Capture group with %s transforms to %s, which is incompatible with %s. " +
+                                "Try changing the capture group to something that doesn't match an existing parameter type.",
+                        parameterTypeRegexp,
+                        paramClass.getTypeName(),
+                        hintClass.getTypeName())
+                );
+            }
+
             parameterTypes.add(parameterType);
```

Comment:
@aslakhellesoy I rewrote the way we handled anonymous parameter types. Now the old comments have disappeared.

You mentioned that you wanted to do validation here as we have all the information available here. However that information (`typeHint`, and `parameterType.type`) is also available to the caller of `Expression.match`.

And when I think about the error message I need much more information. For example in the given situation:

```java
@Given("A transaction of (\d+) ([A-Z]{3})")
public void given_a_transaction_of(Amount a, Currency c)
```
The regex of capture group itself tells me very little. 

I'd like this error message to look something like this:

```
Encountered a problem while executing step: "A transaction of 1000 PTE".

Tried to invoke `given_a_transaction_of(Amount a, Currency c)` with these arguments `(Integer, String)`. 

The arguments types do not match up.

Argument values:
 *  Type: Integer
    Value: 1000
    Regex: \d+
    Name: null
 * Type: String
    Value: PTE
    Regex: [A-Z]{3}
    Name: null

You can resolve this problem by:
 * Change the method parameters to match the arguments.
 * Register the method parameters in the parameter type registry  

Link to docs here:
```